File: packages/sdk-ts/src/resources/grants.ts
import type { HttpClient } from '../http.js';
import { claimsToVerifiedGrant } from '../verify.js';
import { GrantexTokenError } from '../errors.js';
import type {
  Grant,
  ListGrantsParams,
  ListGrantsResponse,
  VerifiedGrant,
  DelegateParams,
  GrantTokenPayload,
} from '../types.js';

export class GrantsClient {
  readonly #http: HttpClient;

  constructor(http: HttpClient) {
    this.#http = http;
  }

  get(grantId: string): Promise<Grant> {
    return this.#http.get<Grant>(`/v1/grants/${grantId}`);
  }

  list(params?: ListGrantsParams): Promise<ListGrantsResponse> {
    const query = buildQuery(params);
    const path = query ? `/v1/grants?${query}` : '/v1/grants';
    return this.#http.get<ListGrantsResponse>(path);
  }

  revoke(grantId: string): Promise<void> {
    return this.#http.delete<void>(`/v1/grants/${grantId}`);
  }

  /**
   * Create a delegated sub-agent grant (SPEC §9).
   */
  delegate(params: DelegateParams): Promise<{ grantToken: string; expiresAt: string; scopes: string[]; grantId: string }> {
    return this.#http.post('/v1/grants/delegate', params);
  }

  /**
   * Verify a grant token via the API (online verification).
   */
  async verify(token: string): Promise<VerifiedGrant> {
    const response = await this.#http.post<{
      active?: boolean;
      reason?: string;
      claims?: GrantTokenPayload;
    }>(
      '/v1/grants/verify',
      { token },
    );
    if (!response.active || !response.claims) {
      throw new GrantexTokenError(
        `Grant token is not active${response.reason ? `: ${response.reason}` : ''}`,
      );
    }
    return claimsToVerifiedGrant(response.claims);
  }
}

function buildQuery(params?: ListGrantsParams): string {
  if (!params) return '';
  const entries = Object.entries(params).filter(
    ([, v]) => v !== undefined && v !== null,
  ) as Array<[string, string | number]>;
  if (entries.length === 0) return '';
  return entries
    .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(String(v))}`)
    .join('&');
}

File: packages/sdk-py/src/grantex/resources/_grants.py
from __future__ import annotations

from typing import Any, List
from urllib.parse import urlencode

from .._errors import GrantexTokenError
from .._http import HttpClient
from .._types import Grant, ListGrantsParams, ListGrantsResponse, VerifiedGrant, DelegateParams
from .._verify import _build_payload, _payload_to_verified_grant


class GrantsClient:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def get(self, grant_id: str) -> Grant:
        data = self._http.get(f"/v1/grants/{grant_id}")
        return Grant.from_dict(data)

    def list(self, params: ListGrantsParams | None = None) -> ListGrantsResponse:
        qs = _build_query(params.to_dict() if params else {})
        path = f"/v1/grants?{qs}" if qs else "/v1/grants"
        data = self._http.get(path)
        return ListGrantsResponse.from_dict(data)

    def revoke(self, grant_id: str) -> None:
        self._http.delete(f"/v1/grants/{grant_id}")

    def delegate(
        self,
        *,
        parent_grant_token: str,
        sub_agent_id: str,
        scopes: List[str],
        expires_in: str | None = None,
    ) -> Any:
        params = DelegateParams(
            parent_grant_token=parent_grant_token,
            sub_agent_id=sub_agent_id,
            scopes=scopes,
            expires_in=expires_in,
        )
        return self._http.post("/v1/grants/delegate", params.to_dict())

    def verify(self, token: str) -> VerifiedGrant:
        response = self._http.post("/v1/grants/verify", {"token": token})
        if not isinstance(response, dict) or not response.get("active") or not response.get("claims"):
            reason = response.get("reason") if isinstance(response, dict) else None
            suffix = f": {reason}" if reason else ""
            raise GrantexTokenError(f"Grant token is not active{suffix}")
        claims: dict[str, Any] = response["claims"]
        return _payload_to_verified_grant(_build_payload(claims))


def _build_query(params: dict[str, object]) -> str:
    filtered = {k: v for k, v in params.items() if v is not None}
    if not filtered:
        return ""
    return urlencode(filtered)
